File: src/mgen/shared/domain_types.py
"""Domain value objects for the Mt Messenger monitoring pipeline."""

from __future__ import annotations

from dataclasses import dataclass

import pandas as pd

# Samples on or before this date are Baseline regardless of source label.
# Per MKE review: baseline monitoring includes all samples through March 2022.
BASELINE_END = pd.Timestamp("2022-03-31")

# Canonical site codes for the Mt Messenger monitoring programme.
# NOTE: "MMA 6" and "MMA 6b" have a space — matches source database values.
VALID_SITES = frozenset(
    {"EM1", "EM2", "EM3", "EM4", "EM5", "EM7", "EM8", "MMA 6", "MMA 6b"}
)

# Sites that lack replicates and use QMCI-sb instead of QMCI
SITES_WITHOUT_REPLICATES = frozenset({"EM1", "EM2", "EM4", "EM8"})

# Valid monitoring period labels (from Sediment sheet first column)
VALID_PERIODS = frozenset({"Baseline", "Construction"})

# Valid season/phase labels used in output Data.xlsx.
# "Routine" is a planned future monitoring phase post-construction.
VALID_SEASONS = frozenset({"Baseline", "Construction", "Routine", "Additional"})


@dataclass(frozen=True)
class Site:
    """A validated monitoring site."""

    code: str

    def __post_init__(self) -> None:
        if self.code not in VALID_SITES:
            msg = f"Unknown site: {self.code!r}. Expected one of {sorted(VALID_SITES)}"
            raise ValueError(msg)

    @property
    def has_replicates(self) -> bool:
        """Whether this site has replicate samples."""
        return self.code not in SITES_WITHOUT_REPLICATES

    @property
    def uses_qmci_sb(self) -> bool:
        """Whether this site uses QMCI-sb instead of QMCI."""
        return self.code in SITES_WITHOUT_REPLICATES

File: src/r/data/domain_types.R
# domain_types.R — shared monitoring constants.

# Samples on or before this date are Baseline regardless of source label.
BASELINE_END <- as.Date("2022-03-31")

# "MMA 6"/"MMA 6b" contain a space — matches source database values.
VALID_SITES <- c("EM1", "EM2", "EM3", "EM4", "EM5", "EM7", "EM8", "MMA 6", "MMA 6b")

# Sites lacking replicates — use QMCI-sb and are excluded from aggregated Macro.
SITES_WITHOUT_REPLICATES <- c("EM1", "EM2", "EM4", "EM8")

# Static per-site attributes from the project site legend. These are fixed
# properties of each monitoring site, not per-sample values.
#   Substrate == "Soft bottom"  -> D-net sampling, MCI-sb/QMCI-sb metrics
#   Substrate == "Hard bottom"  -> Surber sampling (5 replicates averaged), MCI/QMCI
# The soft-bottom set is identical to SITES_WITHOUT_REPLICATES (see test).
# MMA 6 / MMA 6b are omitted: they are not part of the EM site legend.
SITE_LEGEND <- data.frame(
  Site      = c("EM1", "EM2", "EM3", "EM4", "EM5", "EM7", "EM8"),
  Catchment = c("Mangapepeke", "Mangapepeke", "Mangapepeke",
                "Mimi", "Mimi", "Mimi", "Mimi"),
  Substrate = c("Soft bottom", "Soft bottom", "Hard bottom",
                "Soft bottom", "Hard bottom", "Hard bottom", "Soft bottom"),
  Method    = c("D-net", "D-net", "Surber",
                "D-net", "Surber", "Surber", "D-net"),
  IsControl = c(TRUE, FALSE, FALSE, TRUE, FALSE, FALSE, FALSE),
  stringsAsFactors = FALSE
)

VALID_PERIODS <- c("Baseline", "Construction")
VALID_SEASONS <- c("Baseline", "Construction", "Routine", "Additional")
